File: frontend/src/services/riskEngine.ts
export type RiskProduct = 'CNC' | 'MIS'
export type RiskCategory = 'LC' | 'MC' | 'SC' | 'ETF'

export type RiskProfile = {
  id: number
  name: string
  product: RiskProduct

  capital_per_trade: number
  max_positions: number
  max_exposure_pct: number

  risk_per_trade_pct: number
  hard_risk_pct: number

  stop_loss_mandatory: boolean
  stop_reference: 'ATR' | 'FIXED_PCT'
  atr_period: number
  atr_mult_initial_stop: number
  fallback_stop_pct: number
  min_stop_distance_pct: number
  max_stop_distance_pct: number

  daily_loss_pct: number
  hard_daily_loss_pct: number
  max_consecutive_losses: number

  drawdown_mode: 'SETTINGS_BY_CATEGORY'

  force_exit_time?: string | null

  entry_cutoff_time?: string | null
  force_squareoff_time?: string | null
  max_trades_per_day?: number | null
  max_trades_per_symbol_per_day?: number | null
  min_bars_between_trades?: number | null
  cooldown_after_loss_bars?: number | null
  slippage_guard_bps?: number | null
  gap_guard_pct?: number | null
  order_type_policy?: string | null
  leverage_mode?: string | null
  max_effective_leverage?: number | null
  max_margin_used_pct?: number | null

  enabled: boolean
  is_default: boolean

  created_at: string
  updated_at: string
}

export type RiskProfileCreate = Omit<RiskProfile, 'id' | 'created_at' | 'updated_at'>
export type RiskProfileUpdate = Partial<RiskProfileCreate>

export type DrawdownThreshold = {
  id: number
  user_id?: number | null
  product: RiskProduct
  category: RiskCategory
  caution_pct: number
  defense_pct: number
  hard_stop_pct: number
  created_at: string
  updated_at: string
}

export type DrawdownThresholdUpsert = Pick<
  DrawdownThreshold,
  'product' | 'category' | 'caution_pct' | 'defense_pct' | 'hard_stop_pct'
>

export type SymbolRiskCategory = {
  id: number
  user_id?: number | null
  broker_name: string
  symbol: string
  exchange: string
  risk_category: RiskCategory
  created_at: string
  updated_at: string
}

export type SymbolRiskCategoryUpsert = Pick<
  SymbolRiskCategory,
  'broker_name' | 'symbol' | 'exchange' | 'risk_category'
>

export type AlertDecisionLogRow = {
  id: number
  created_at: string
  user_id?: number | null
  alert_id?: number | null
  order_id?: number | null

  source: string
  strategy_ref?: string | null
  symbol?: string | null
  exchange?: string | null
  side?: string | null
  trigger_price?: number | null

  product_hint?: string | null
  resolved_product?: string | null
  risk_profile_id?: number | null
  risk_category?: string | null
  drawdown_pct?: number | null
  drawdown_state?: string | null

  decision: string
  reasons_json: string
  details_json: string
}

export type RiskEngineV2Enabled = {
  enabled: boolean
  source: 'db' | 'env_default' | 'db_invalid'
  updated_at?: string | null
}

async function readTextSafe(res: Response): Promise<string> {
  try {
    return await res.text()
  } catch {
    return ''
  }
}

export async function fetchRiskEngineV2Enabled(): Promise<RiskEngineV2Enabled> {
  const res = await fetch('/api/risk-engine/v2-enabled', { cache: 'no-store' })
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(
      `Failed to load profile engine status (${res.status})${body ? `: ${body}` : ''}`,
    )
  }
  return (await res.json()) as RiskEngineV2Enabled
}

export async function updateRiskEngineV2Enabled(enabled: boolean): Promise<RiskEngineV2Enabled> {
  const res = await fetch('/api/risk-engine/v2-enabled', {
    method: 'PUT',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ enabled: Boolean(enabled) }),
  })
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(
      `Failed to update profile engine status (${res.status})${body ? `: ${body}` : ''}`,
    )
  }
  return (await res.json()) as RiskEngineV2Enabled
}

export async function fetchRiskProfiles(): Promise<RiskProfile[]> {
  const res = await fetch('/api/risk-engine/risk-profiles')
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(`Failed to load risk profiles (${res.status})${body ? `: ${body}` : ''}`)
  }
  return (await res.json()) as RiskProfile[]
}

export async function createRiskProfile(payload: RiskProfileCreate): Promise<RiskProfile> {
  const res = await fetch('/api/risk-engine/risk-profiles', {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(`Failed to create risk profile (${res.status})${body ? `: ${body}` : ''}`)
  }
  return (await res.json()) as RiskProfile
}

export async function updateRiskProfile(
  profileId: number,
  payload: RiskProfileUpdate,
): Promise<RiskProfile> {
  const res = await fetch(`/api/risk-engine/risk-profiles/${profileId}`, {
    method: 'PUT',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(`Failed to update risk profile (${res.status})${body ? `: ${body}` : ''}`)
  }
  return (await res.json()) as RiskProfile
}

export async function deleteRiskProfile(profileId: number): Promise<void> {
  const res = await fetch(`/api/risk-engine/risk-profiles/${profileId}`, {
    method: 'DELETE',
  })
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(`Failed to delete risk profile (${res.status})${body ? `: ${body}` : ''}`)
  }
}

export async function fetchDrawdownThresholds(): Promise<DrawdownThreshold[]> {
  const res = await fetch('/api/risk-engine/drawdown-thresholds')
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(
      `Failed to load drawdown thresholds (${res.status})${body ? `: ${body}` : ''}`,
    )
  }
  return (await res.json()) as DrawdownThreshold[]
}

export async function upsertDrawdownThresholds(
  payload: DrawdownThresholdUpsert[],
): Promise<DrawdownThreshold[]> {
  const res = await fetch('/api/risk-engine/drawdown-thresholds', {
    method: 'PUT',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(
      `Failed to save drawdown thresholds (${res.status})${body ? `: ${body}` : ''}`,
    )
  }
  return (await res.json()) as DrawdownThreshold[]
}

export async function fetchSymbolCategories(brokerName = 'zerodha'): Promise<SymbolRiskCategory[]> {
  const url = new URL('/api/risk-engine/symbol-categories', window.location.origin)
  url.searchParams.set('broker_name', brokerName)
  const res = await fetch(url.toString(), { cache: 'no-store' })
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(`Failed to load symbol categories (${res.status})${body ? `: ${body}` : ''}`)
  }
  return (await res.json()) as SymbolRiskCategory[]
}

export async function upsertSymbolCategory(
  payload: SymbolRiskCategoryUpsert,
): Promise<SymbolRiskCategory> {
  const res = await fetch('/api/risk-engine/symbol-categories', {
    method: 'PUT',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(`Failed to save symbol category (${res.status})${body ? `: ${body}` : ''}`)
  }
  return (await res.json()) as SymbolRiskCategory
}

export async function deleteSymbolCategory(params: {
  broker_name?: string
  exchange?: string
  symbol: string
}): Promise<void> {
  const url = new URL('/api/risk-engine/symbol-categories', window.location.origin)
  url.searchParams.set('symbol', params.symbol)
  url.searchParams.set('broker_name', params.broker_name ?? 'zerodha')
  url.searchParams.set('exchange', params.exchange ?? 'NSE')
  const res = await fetch(url.toString(), { method: 'DELETE' })
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(
      `Failed to delete symbol category (${res.status})${body ? `: ${body}` : ''}`,
    )
  }
}

export async function bulkUpsertSymbolCategories(
  payload: SymbolRiskCategoryUpsert[],
): Promise<SymbolRiskCategory[]> {
  const res = await fetch('/api/risk-engine/symbol-categories/bulk', {
    method: 'PUT',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(
      `Failed to bulk save symbol categories (${res.status})${body ? `: ${body}` : ''}`,
    )
  }
  return (await res.json()) as SymbolRiskCategory[]
}

export async function fetchAlertDecisionLog(limit = 100): Promise<AlertDecisionLogRow[]> {
  const url = new URL('/api/risk-engine/decision-log', window.location.origin)
  url.searchParams.set('limit', String(limit))
  const res = await fetch(url.toString(), { cache: 'no-store' })
  if (!res.ok) {
    const body = await readTextSafe(res)
    throw new Error(`Failed to load decision log (${res.status})${body ? `: ${body}` : ''}`)
  }
  return (await res.json()) as AlertDecisionLogRow[]
}

File: backend/app/schemas/risk_engine.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


RiskProduct = Literal["CNC", "MIS"]
RiskCategory = Literal["LC", "MC", "SC", "ETF"]
DrawdownState = Literal["NORMAL", "CAUTION", "DEFENSE", "HALT"]


RiskEngineV2FlagSource = Literal["db", "env_default", "db_invalid"]


class RiskEngineV2FlagRead(BaseModel):
    enabled: bool
    source: RiskEngineV2FlagSource
    updated_at: datetime | None = None


class RiskEngineV2FlagUpdate(BaseModel):
    enabled: bool


class RiskProfileBase(BaseModel):
    name: str
    product: RiskProduct

    capital_per_trade: float = 0.0
    max_positions: int = 0
    max_exposure_pct: float = 0.0

    risk_per_trade_pct: float = 0.0
    hard_risk_pct: float = 0.0

    daily_loss_pct: float = 0.0
    hard_daily_loss_pct: float = 0.0
    max_consecutive_losses: int = 0

    drawdown_mode: Literal["SETTINGS_BY_CATEGORY"] = "SETTINGS_BY_CATEGORY"

    force_exit_time: str | None = None

    entry_cutoff_time: str | None = None
    force_squareoff_time: str | None = None
    max_trades_per_day: int | None = None
    max_trades_per_symbol_per_day: int | None = None
    min_bars_between_trades: int | None = None
    cooldown_after_loss_bars: int | None = None
    slippage_guard_bps: float | None = None
    gap_guard_pct: float | None = None
    order_type_policy: str | None = None
    leverage_mode: str | None = None
    max_effective_leverage: float | None = None
    max_margin_used_pct: float | None = None

    enabled: bool = True
    is_default: bool = False


class RiskProfileCreate(RiskProfileBase):
    pass


class RiskProfileUpdate(BaseModel):
    name: Optional[str] = None
    product: Optional[RiskProduct] = None

    capital_per_trade: Optional[float] = None
    max_positions: Optional[int] = None
    max_exposure_pct: Optional[float] = None

    risk_per_trade_pct: Optional[float] = None
    hard_risk_pct: Optional[float] = None

    daily_loss_pct: Optional[float] = None
    hard_daily_loss_pct: Optional[float] = None
    max_consecutive_losses: Optional[int] = None

    drawdown_mode: Optional[Literal["SETTINGS_BY_CATEGORY"]] = None

    force_exit_time: Optional[str | None] = None

    entry_cutoff_time: Optional[str | None] = None
    force_squareoff_time: Optional[str | None] = None
    max_trades_per_day: Optional[int | None] = None
    max_trades_per_symbol_per_day: Optional[int | None] = None
    min_bars_between_trades: Optional[int | None] = None
    cooldown_after_loss_bars: Optional[int | None] = None
    slippage_guard_bps: Optional[float | None] = None
    gap_guard_pct: Optional[float | None] = None
    order_type_policy: Optional[str | None] = None
    leverage_mode: Optional[str | None] = None
    max_effective_leverage: Optional[float | None] = None
    max_margin_used_pct: Optional[float | None] = None

    enabled: Optional[bool] = None
    is_default: Optional[bool] = None


class RiskProfileRead(RiskProfileBase):
    id: int
    created_at: datetime
    updated_at: datetime


class SymbolRiskCategoryUpsert(BaseModel):
    broker_name: str = "zerodha"
    symbol: str
    exchange: str = "NSE"
    risk_category: RiskCategory


class SymbolRiskCategoryRead(SymbolRiskCategoryUpsert):
    id: int
    user_id: int | None
    created_at: datetime
    updated_at: datetime


class DrawdownThresholdUpsert(BaseModel):
    product: RiskProduct
    category: RiskCategory
    caution_pct: float = Field(ge=0)
    defense_pct: float = Field(ge=0)
    hard_stop_pct: float = Field(ge=0)


class DrawdownThresholdRead(DrawdownThresholdUpsert):
    id: int
    user_id: int | None
    created_at: datetime
    updated_at: datetime


class EquitySnapshotRead(BaseModel):
    id: int
    user_id: int | None
    as_of_date: date
    equity: float
    peak_equity: float
    drawdown_pct: float
    created_at: datetime
    updated_at: datetime


class AlertDecisionLogRead(BaseModel):
    id: int
    created_at: datetime
    user_id: int | None
    alert_id: int | None
    order_id: int | None

    source: str
    strategy_ref: str | None
    symbol: str | None
    exchange: str | None
    side: str | None
    trigger_price: float | None

    product_hint: str | None
    resolved_product: str | None
    risk_profile_id: int | None
    risk_category: str | None
    drawdown_pct: float | None
    drawdown_state: str | None

    decision: str
    reasons_json: str = "[]"
    details_json: str = "{}"


__all__ = [
    "AlertDecisionLogRead",
    "DrawdownState",
    "DrawdownThresholdRead",
    "DrawdownThresholdUpsert",
    "EquitySnapshotRead",
    "RiskEngineV2FlagRead",
    "RiskEngineV2FlagUpdate",
    "RiskCategory",
    "RiskProduct",
    "RiskProfileCreate",
    "RiskProfileRead",
    "RiskProfileUpdate",
    "SymbolRiskCategoryRead",
    "SymbolRiskCategoryUpsert",
]
